src/segtree.rs

```diff
@@ -1 +1,273 @@
+use crate::internal_bit::ceil_pow2;
+use crate::internal_type_traits::Integral;
+use std::cmp::max;
+use std::marker::PhantomData;
 
+// TODO Should I split monoid-related traits to another module?
+pub trait Monoid {
+    type S: Clone;
+    fn identity() -> Self::S;
+    fn binary_operation(a: Self::S, b: Self::S) -> Self::S;
```

Comment:
If we don't require `S: Copy`, why not change `binary_operation(S, S) -> S` to `〃(&S, &S) -> S`?

Reply:
how about this

```rust
pub enum Max<S> {
    __Void(Infallible, PhantomData<fn() -> S>),
}

impl<S: Zero + Copy + Ord + Bounded> Monoid for Max<S> {
    type S = S;

    fn identity() -> S {
        S::zero()
    }

    fn op(lhs: &S, rhs: &S) -> S {
        *lhs + *rhs
    }
}

// Like `num_traits::Zero`.
// https://docs.rs/num-traits/0.2.12/num_traits/identities/trait.Zero.html
pub trait Zero: Sized + Add<Output = Self> {
    fn zero() -> Self;
}

// Like `num_traits::Bounded`.
// https://docs.rs/num-traits/0.2.12/num_traits/bounds/trait.Bounded.html
pub trait Bounded {
    fn min_value() -> Self;
    fn max_value() -> Self;
}

macro_rules! impl_for_primitive_integers {
    ($($ty:ty),*) => {
        $(
            impl Zero for $ty {
                fn zero() -> Self {
                    0
                }
            }

            impl Bounded for $ty {
                fn min_value() -> Self {
                    Self::min_value()
                }

                fn max_value() -> Self {
                    Self::max_value()
                }
            }
        )*
    }
}

impl_for_primitive_integers!(i8, i16, i32, i64, i128, isize, u8, u16, u32, u64, u128, usize);
```

Reply:
The design of `Zero` looks good to me.  (Is it redundant if we use the name like `AdditiveIdentity` or something?)

Changing the signature to take reference instead of value may be good, but then the users should not forget writing `&` for each argument in trait implementation right?  Doesn't it bother?  (Yes, it's only a couple-of-bit change)
Also I didn't understand why the binary operations are inlined when references are used.  Is there any documentation about it?

Thirdly, why do you always prefer `enum Max<S> { __Void(Infallible, PhantomData<fn() -> S>) }` to `struct Max<S>(PhantomData<fn() -> S>)`?  I want to know the reason if any.
(BTW I've learned that we should use `PhantomData<fn() -> S>` rather than `PhantomData<S>` from [this article](https://qnighy.hatenablog.com/entry/2018/01/14/220000), which is very unnatural to me)

Reply:
1. `num_traits::Zero` is defined as

    > an additive identity
    
    and users (particularly who are familiar with Rust) will expect that our `Zero` is so. I think that's enough.

2.  I hadn't thought of that. I thought you were worrying about performance. If we will allow non-`Copy` large heavy-to-copy data, then we shouldn't force users to `clone` them. Actually we can take both of `S` and `&S` by setting type of the arguments to `impl Borrow<S>` or something, but I don't think this is idiomatic.

    ```rust
    use std::borrow::Borrow;

    fn pass(_: impl Borrow<i32>) {}

    pass(0);
    pass(&0);
    pass(&mut 0);
    ```

    And

    > why the binary operations are inlined when references are used

    Rust inlines functions with small body as other general-purpose programming languages do.

    [The Rust Reference](https://doc.rust-lang.org/reference/attributes/codegen.html#the-inline-attribute)

3. Because the instances of marker types don't need to exist. If a marker type does not take generic parameter, this kind of definition is commonly used.

    ```rust
    enum Marker {}
    ```

    But I'm not sure adding `Infallible` is idiomatic when we need to include `PhantomData`.

Reply:
1. Got it.  I'll use the simple name `Zero`, and additionally `One` for product identity.
2. I understood.  I'll change the signature later (and that of lazy segtree as well).
3. I see.  If the purpose is to prevent creating instance, then isn't `struct Max<S>(Infallible, PhantomData<fn() -> S>)` enough?

Reply:
Yes, probably it's enough.

Reply:
Should we change the types of `f` in `max_right` and `min_left` too so that it takes reference as an argument?

Reply:
Yes, I think we should change them for consistency.

Reply:
Fixed.